node/src/components/small_network/tasks.rs

```diff
@@ -353,6 +362,33 @@ where
             return Err(ConnectionError::WrongNetwork(network_name));
         }
 
+        // If there is a version mismatch, we treat it as a connection error. We do not ban peers
+        // for this error, but instead rely on exponential backoff, as bans would result in issues
+        // during upgrades where nodes may have a legitimate reason for differing versions.
+        //
+        // Since we are not using SemVer for versioning, we cannot make any assumptions about
+        // compatibility, so we allow only exact version matches.
+        if context.reject_incompatible_versions
+            && protocol_version != context.chain_info.protocol_version
+        {
+            if let Some(threshold) = context.tarpit_version_threshold {
+                if protocol_version <= threshold {
+                    let mut rng = crate::new_rng();
+
+                    if rng.gen_bool(context.tarpit_chance as f64) {
+                        // If tarpitting is enabled, we hold open the connection for a specific
+                        // amount of time, to reduce load on other nodes and keep them from
+                        // reconnecting.
+                        info!(duration=?context.tarpit_duration, "tarpitting node");
+                        tokio::time::sleep(context.tarpit_duration).await;
+                    } else {
+                        debug!(p = context.tarpit_chance, "randomly not tarpitting node");
```

Comment:
I'd add the `randomly not` to the other message, too, or remove it from this one.

Reply:
But we are tarpitting it on one instance and not in the other? I'm not sure I understand this suggestion :)

Reply:
Agree with Andreas - let's keep the messages consistent.

Reply:
Ah, makes more sense now. Got it :)